subprojects/snapshots/src/main/java/org/gradle/internal/fingerprint/impl/NormalizedPathFingerprintCompareStrategy.java

```diff
@@ -56,61 +55,138 @@ private NormalizedPathFingerprintCompareStrategy() {
      *         For those only in the previous fingerprint collection it checks if some entry with the same normalized path is in the current collection.
      *         If it is, file is reported as modified, if not as removed.
      *     </li>
-     *     <li>Finally, if {@code includeAdded} is {@code true}, the remaining fingerprints which are only in the current collection are reported as added.</li>
+     *     <li>Finally, {@code includeAdded} is always {@code true}, meaning that the remaining fingerprints which are only in the current collection are reported as added.</li>
      * </ul>
      */
     @Override
-    protected boolean doVisitChangesSince(ChangeVisitor visitor, Map<String, FileSystemLocationFingerprint> currentFingerprints, Map<String, FileSystemLocationFingerprint> previousFingerprints, String propertyTitle, boolean includeAdded) {
-        ListMultimap<FileSystemLocationFingerprint, FilePathWithType> unaccountedForPreviousFiles = MultimapBuilder.hashKeys(previousFingerprints.size()).linkedListValues().build();
-        ListMultimap<String, FilePathWithType> addedFilesByNormalizedPath = MultimapBuilder.linkedHashKeys().linkedListValues().build();
-        for (Map.Entry<String, FileSystemLocationFingerprint> entry : previousFingerprints.entrySet()) {
+    protected boolean doVisitChangesSince(
+        ChangeVisitor visitor,
+        Map<String, FileSystemLocationFingerprint> currentFingerprints,
+        Map<String, FileSystemLocationFingerprint> previousFingerprints,
+        String propertyTitle,
+        boolean includeAdded
+    ) {
+        Preconditions.checkArgument(includeAdded);
+        return doVisitChangesSince(visitor, currentFingerprints, previousFingerprints, propertyTitle);
+    }
+
+    private boolean doVisitChangesSince(
+        ChangeVisitor visitor,
+        Map<String, FileSystemLocationFingerprint> currentFingerprints,
+        Map<String, FileSystemLocationFingerprint> previousFingerprints,
+        String propertyTitle
+    ) {
+        ListMultimap<FileSystemLocationFingerprint, FilePathWithType> unaccountedForPreviousFiles = getUnaccountedForPreviousFingerprints(previousFingerprints, currentFingerprints);
+        ListMultimap<String, FilePathWithType> addedFilesByNormalizedPath = getAddedFilesByNormalizedPath(currentFingerprints, unaccountedForPreviousFiles, previousFingerprints);
+
+        Iterator<Entry<FileSystemLocationFingerprint, FilePathWithType>> iterator = unaccountedForPreviousFiles.entries().stream().sorted(comparingByKey()).iterator();
+        while (iterator.hasNext()) {
+            Entry<FileSystemLocationFingerprint, FilePathWithType> entry = iterator.next();
```

Comment:
Why create an iterator? Looks like a `.forEach()` would be the natural fit here.

Reply:
How do I return from a foreach early?